### ISSUE TITLE
AccuracyCalculator doesn't work correctly for mAP@k when "k" is specified

### ISSUE THREAD

**Issue by vortexdev** [KevinMusgrave/pytorch-metric-learning#240]
I recently switched from my implementation of the Mean Average Precision (mAP@k) to AccuracyCalculator, but the results differ from mine. I have also compared them to other implementations and it looks like there is a bug in the AccuracyCalculator (in my case this results in mAP@10=0.412 instead of 0.515 calculated by other implementations, which is quite a big difference).

Unfortunately, I haven't yet found the exact location of the bug in the source code, but we can reproduce it using existing tests: https://github.com/KevinMusgrave/pytorch-metric-learning/blob/72d7496c419b4ff89d4b6443cc8a60c8959ee7d9/tests/utils/test_calculate_accuracies.py.

**Steps to reproduce:**
Existing tests don't verify the cases when the "k" argument is specified. Let's specify **k = 5** in the test:
```
AC = accuracy_calculator.AccuracyCalculator(
        exclude=("NMI", "AMI"), avg_of_avgs=avg_of_avgs, k=5
)
```

**Expected results:**
The test should fail.
Explanation (below code is taken from the test):
```
query_labels = np.array([1, 1, 2, 3, 4])
knn_labels1 = np.array(
            [
                [0, 1, 1, 2, 2],
                [1, 0, 1, 1, 3],
                [4, 4, 4, 4, 2],
                [3, 1, 3, 1, 3],
                [0, 0, 4, 2, 2],
            ]
        )
label_counts1 = {1: 3, 2: 5, 3: 4, 4: 5}
```
For instance, let's take the 4th query_label (`3`), and compare it with the corresponding knn_labels `[3, 1, 3, 1, 3]`.
When `k` is not specified it should be determined automatically (according to the documentation), which is equal to label_count, AFAIU. Thus `k = 4` and `acc3 = (1 + 2.0 / 3) / 4` (code from the test).
If `k = 5`, `acc3` should be equal to `(1 + 2.0/3 + 3.0/5) / 4`, which means that the test should fail when we change `k`.

**Actual results:**
The test does not fail.

Please correct me if I am missing something.

**Comment by KevinMusgrave** [comment]
This is a documentation fail, and an implementation fail depending on how you view it.

When you set k, it'll still do mAP@R, where R is equal to ```min(label_count, k)```. In other words, if you set k to 10, then for a class with a label count of 100, it will do mAP@10, but for classes with a label count of 5, it'll do mAP@5 etc.

The idea was to allow the user to keep using mAP@R, but limit the k-nn search, since that can be slow for large k.

I can make it so that mAP@R switches to mAP@k when k is specified, or I can make a separate mAP function which always uses a fixed k value.

**Comment by KevinMusgrave** [comment]
I have added ```"mean_average_precision"``` as one of the default metrics in AccuracyCalculator. It's available in v0.9.95.dev1. 

```
pip install pytorch-metric-learning==0.9.95.dev1
```

It should match other implementations.

**Comment by vortexdev** [comment]
Hi Kevin,

Now I see that the name of the metric was `mean_average_precision_r`, looks like 'k' param confused me. I think adding an additional `mean_average_precision` is a good option.

Thank you!
